docs/source/developers/cpp/compute.rst

```diff
@@ -0,0 +1,171 @@
+.. Licensed to the Apache Software Foundation (ASF) under one
+.. or more contributor license agreements.  See the NOTICE file
+.. distributed with this work for additional information
+.. regarding copyright ownership.  The ASF licenses this file
+.. to you under the Apache License, Version 2.0 (the
+.. "License"); you may not use this file except in compliance
+.. with the License.  You may obtain a copy of the License at
+
+..   http://www.apache.org/licenses/LICENSE-2.0
+
+.. Unless required by applicable law or agreed to in writing,
+.. software distributed under the License is distributed on an
+.. "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+.. KIND, either express or implied.  See the License for the
+.. specific language governing permissions and limitations
+.. under the License.
+
+.. highlight:: console
+.. _development-cpp-compute:
+
+============================
+Developing Arrow C++ Compute
+============================
+
+This section provides information for developers of Arrow C++ Compute module.
+
+Row Table
+=========
+
+The row table in Arrow represents data stored in row-major format. This format
+is particularly useful for scenarios involving random access to individual rows
+and where all columns are frequently accessed together. It is especially
+advantageous for hash-table keys and facilitates efficient operations such as
+grouping and hash joins by optimizing memory access patterns and data locality.
+
+Fixed-length vs. Varying-length
+-------------------------------
+
+A key property of the row table is whether it is fixed-length or varying-length.
+A fixed-length row table contains only fixed-length columns, while a
+varying-length row table includes at least one varying-length column. This
+distinction determines how data is stored and accessed in the row table.
```

Comment:
I think this is confusingly worded or maybe I misunderstand. "varying-length column" implies to me the number of records/rows is what varies but I think, in this API, we're allowing the number of fields in a row to vary. i.e., each row is either an n-tuple where n is fixed or where n can vary. The text below this also uses language which maybe could be clearer. Such as,

> column in that row is null

could be

> field in that row is null

What do you think?

Reply:
Yes, I agree this can be misleading. I tent to keep using this wording because they are the same as the names of variable/function/comment in the codebase, and IMO it can build direct mapping from doc to code. To disambiguate , how about `fixed-length column` -> `fixed-length type column (hereinafter "fixed-length column")` and `varying-length column` -> `varying-length length column (hereinafter "varying-length column")`.

Reply:
Oh, and to be clear, a row table is fully schema-ed. That is, each row contains the same number of columns/fields, in the same types and order as specified in the schema.

Reply:
The text below this also uses language which maybe could be clearer. Such as,
> 
> > column in that row is null
> 
> could be
> 
> > field in that row is null
> 
> What do you think?

I think in the context of the row table, "column" and "field" can be used interchangeably? Will it feel less confusing after the explanation I made above?

Reply:
> Oh, and to be clear, a row table is fully schema-ed. That is, each row contains the same number of columns/fields, in the same types and order as specified in the schema.

Maybe I should add this in the doc too.

Reply:
Ah, I see your point about matching the language here with what's already in the docstrings. Maybe it's not so confusing to people as you have it then. I don't think clarifying (like "fixed-length type column (hereinafter "fixed-length column")") is needed then.

Reply:
Hi @amoeba , I've restructured the doc a bit, by introducing a "Metadata" section that includes some description about the schema-ful nature of a row table, and fixed-length/varying-length columns derived from the schema. Also the last section of alignment is also moved under it. Hope this can help on the disambiguation.